kolibri/core/discovery/utils/network/search.py

```diff
@@ -130,31 +145,75 @@ def remove_service(self, zeroconf, type, name):
+    instance["data"]["channels"] = network_client.get(endpoint_path).json()
+
+
+def find_peer_instances():
     """Retrieve a list of dicts with information about the discovered Kolibri instances on the local network,
     filtering out those that can't be accessed at the specified port (via attempting to open a socket)."""
     if not ZEROCONF_STATE["listener"]:
         initialize_zeroconf_listener()
         time.sleep(3)
-    instances = []
-    for instance in ZEROCONF_STATE["listener"].instances.values():
-        if instance["local"] and not include_local:
-            continue
-        if not _is_port_open(instance["ip"], instance["port"], timeout=timeout):
-            continue
-        instance["self"] = (
-            ZEROCONF_STATE["service"] and ZEROCONF_STATE["service"].id == instance["id"]
-        )
-        instances.append(instance)
-    return instances
+
+    return list(ZEROCONF_STATE["listener"].instances.values())
+
+
+def get_available_instances(timeout=2, include_local=True):
+    """Find peer Kolibri instances, check their availability then add them to the database.
+    Returns either a `DynamicNetworkLocation` queryset or a list of DynamicNetworkLocation objects."""
+
+    discoveries_are_fresh = cache.get(ZEROCONF_DISCOVERIES_ARE_FRESH)
+
+    if discoveries_are_fresh:
+        # a queryset pointing at already discovered addresses
+        return DynamicNetworkLocation.objects
+    else:
+
+        for instance in find_peer_instances():
+            if instance["local"] and not include_local:
+                continue
+
+            if instance["self"]:
+                continue
+
+            try:
+                (
+                    network_location,
+                    created,
+                ) = DynamicNetworkLocation.objects.update_or_create(
+                    dict(base_url=instance.get("base_url")),
+                    id=instance.get("data").get("instance_id"),
+                )
+
+            except NetworkLocationNotFound:
+                logger.info(
+                    "The device with id %s could no longer be reached" % instance["id"]
+                )
+
+        cache.set(ZEROCONF_DISCOVERIES_ARE_FRESH, True, ZEROCONF_MIN_ALLOWED_REFRESH)
+        return DynamicNetworkLocation.objects
 
 
 def register_zeroconf_service(port, id):
+    short_id = id[:4]
+    DynamicNetworkLocation.objects.purge()  # cleanup old dynamic network locations
+    cache.set(ZEROCONF_SERVICE_ID, short_id)
     if ZEROCONF_STATE["service"] is not None:
         unregister_zeroconf_service()
-    logger.info("Registering ourselves to zeroconf network with id '%s'..." % id)
-    data = {"version": kolibri.VERSION}
-    ZEROCONF_STATE["service"] = KolibriZeroconfService(id=id, port=port, data=data)
+    logger.info("Registering ourselves to zeroconf network with id '%s'..." % short_id)
+    data = {
+        "version": kolibri.VERSION,
+        "instance_id": id,
```

Comment:
Here we're passing the full `instance_id` through zeroconf.  It will be used as the id for the `DynamicNetworkLocation` object, ultimately to make it easier to reference in the content import wizard.